src/main/java/org/wildfly/security/sasl/util/AuthenticationTimeoutSaslServerFactory.java

```diff
@@ -0,0 +1,146 @@
+import javax.security.auth.callback.NameCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.sasl.SaslServer;
+import javax.security.sasl.SaslServerFactory;
+import javax.security.sasl.SaslException;
+
+import org.wildfly.common.Assert;
+import org.wildfly.security.sasl.WildFlySasl;
+
+/**
+ * A {@link SaslServerFactory} which adds authentication timeout functionality to a delegate {@code SaslServerFactory}.
+ * <p>
+ * This {@link SaslServerFactory} must be outside of the {@link AuthenticationCompleteCallbackSaslServerFactory} in the
+ * chain of SASL server factories.
+ *
+ * @author <a href="mailto:fjuma@redhat.com">Farah Juma</a>
+ */
+public final class AuthenticationTimeoutSaslServerFactory extends AbstractDelegatingSaslServerFactory {
+
+    /**
+     * The default amount of time, in seconds, after which an authentication attempt should time out.
+     */
+    public static final long DEFAULT_TIMEOUT = 300;
+
+    private final ScheduledExecutorService scheduledExecutorService;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param delegate the delegate {@code SaslServerFactory}
+     * @param scheduledExecutorService the scheduled executor to use to handle authentication timeout tasks
+     */
+    public AuthenticationTimeoutSaslServerFactory(final SaslServerFactory delegate, final ScheduledExecutorService scheduledExecutorService) {
+        super(delegate);
+        Assert.checkNotNullParam("scheduledExecutorService", scheduledExecutorService);
+        this.scheduledExecutorService = scheduledExecutorService;
+    }
+
+    @Override
+    public SaslServer createSaslServer(final String mechanism, final String protocol, final String serverName, final Map<String, ?> props, final CallbackHandler cbh) throws SaslException {
+        final long timeout;
+        if (props.containsKey(WildFlySasl.AUTHENTICATION_TIMEOUT)) {
+            timeout = Long.parseLong((String) props.get(WildFlySasl.AUTHENTICATION_TIMEOUT));
+        } else {
+            timeout = DEFAULT_TIMEOUT;
+        }
+        final NameAssignedCallbackHandler nameAssignedCallbackHandler = new NameAssignedCallbackHandler(cbh);
+        final SaslServer delegateSaslServer = delegate.createSaslServer(mechanism, protocol, serverName, props, nameAssignedCallbackHandler);
+        return delegateSaslServer == null ? null : new AbstractDelegatingSaslServer(delegateSaslServer) {
+            private final AtomicBoolean complete = new AtomicBoolean();
+            private final AtomicBoolean nameAssigned = new AtomicBoolean();
+            private volatile ScheduledFuture<Void> timeoutTask;
+
+            @Override
+            public byte[] evaluateResponse(final byte[] response) throws SaslException {
+                try {
+                    final byte[] challenge = delegate.evaluateResponse(response);
+                    if (nameAssignedCallbackHandler.isNameAssigned() && nameAssigned.compareAndSet(false, true)) {
```

Comment:
I think we want the authentication timeout to apply no matter whether we have assigned a name; it should probably start counting as soon as the first response is evaluated.